File: 589-nary-preorder/cpp/main.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val) {
        val = _val;
    }

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
 public:
  vector<int> preorder(Node* root) {
    if (root == NULL) {
      return {};
    }
    stack<Node*> stk;
    vector<int> res;
    stk.push(root);

    while (!stk.empty()) {
      auto node = stk.top();
      stk.pop();

      for (auto i = node->children.rbegin(); i != node->children.rend(); ++i) {
        stk.push(*i);
      }

      res.push_back(node->val);
    }

    return res;
  }
};

File: 589-nary-preorder/go/main.go
package main

/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Children []*Node
 * }
 */

func preorder(root *Node) []int {
	if root == nil {
		return []int{}
	}

	stack := []*Node{root}
	res := []int{}
	for len(stack) != 0 {
		n := stack[len(stack)-1]
		stack = stack[:len(stack)-1]

		res = append(res, n.Val)

		for i := len(n.Children) - 1; i >= 0; i-- {
			stack = append(stack, n.Children[i])
		}
	}

	return res
}
